modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h

```diff
@@ -0,0 +1,41 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef CRACKFRONTPOINTSPROVIDER_H
+#define CRACKFRONTPOINTSPROVIDER_H
+
+#include "GeneralUserObject.h"
+
+class CrackFrontPointsProvider;
+
+template <>
+InputParameters validParams<CrackFrontPointsProvider>();
+
+/**
+ * Base class for crack front points provider
+ */
+class CrackFrontPointsProvider : public GeneralUserObject
+{
+public:
+  CrackFrontPointsProvider(const InputParameters & parameters);
+
+  virtual void initialSetup() override{};
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override{};
```

Comment:
You are aren't using any part of the GeneralUserObject interface. Should this just be a new base class instead?

Reply:
CrackFrontDefinition is in TensorMechanics and we want it to use with XFEMCutterUO which exists in XFEM. Now XFEM module depends on TensorMechanics module and we think to have a base class in TensorMechanics and let XFEMCutterUO to derive from it would resolve a dependence issue. This new base class should be UO such that we can use getUserObject interface. Do you have better suggestion?

Reply:
The first part of this comment doesn't make sense to me. Are you trying to avoid depending on tensor mechanics? If you want to make that dependency permanent we can in the modules.mk file, then there's no problem with that.

I don't see any reason to use the getUserObject interface. You are just pulling this object into other objects but there's not execution dependencies. You're just using this object. From that standpoint you could just create a new base class and include the header any place it needs to be used and pull it in. I just don't see what the user object or the interface buys you since you are bypassing 100% of the interface anyway.

Reply:
We want XFEM depend on TM. But the question is that CrackFrontDefinition implemented in TM need to use the XFEMCutterUO class which is in XFEM. This seems to have TM depend on XFEM. We do not have this dependence issue and I am not sure what is the best to do it.

Reply:
Oh I see, Hmm... technically that would work on the compilation side of things. I'm not sure I really want it logically though. It would mean that people get XFEM if they use TM too.

Reply:
We designed it this way specifically because we did not want to force people using TM to have to pull in XFEM.

Reply:
At beginning, all the XFEMCutterObjects were not UO and we ended up with creating them and providing their parameters in XFEMAction, which is a bit cumbersome. And then we decide to have XFEMCutterObject be generalUO and we could provide parameters easier and add/remove them in the input file. The XFEMCutterObject does not really override those basic virtual functions. The new CrackFrontPointsProvider UO here just provides a base class to let XFEMCutterObject derive from. And I could also use getUserObject in the CrackFrontDefinition class to get it.

Reply:
I talked this over with @permcody.  The main reason for using UserObjects without using any of that class's interface is that it is convenient to create and parse the parameters for these objects. Until we have something that offers us that for a more stripped down object, let's keep using UserObjects for this purpose.